src/redisearch_rs/buffer/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    /// Returns the remaining capacity of the buffer (capacity - length).
+    pub fn remaining_capacity(&self) -> usize {
+        self.capacity - self.len
+    }
+
+    /// Advance the buffer by `n` bytes.
+    ///
+    /// This increases the buffer's length by `n` bytes, effectively marking
+    /// more of the buffer as "initialized" without actually writing any data.
+    /// Typically used after directly writing to the buffer's memory.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `n` exceeds the remaining capacity of the buffer.
+    pub fn advance(&mut self, n: usize) {
+        assert!(n <= self.remaining_capacity());
+        self.len += n;
+    }
+}
+
+/// Redefines the `BufferWriter` struct from `buffer.h`
+///
+/// A writer for appending data to a `Buffer`. Provides a mechanism to write bytes to a `Buffer`
+/// while tracking the current write position with a cursor.
+///
+/// # Safety Invariants
+///
+/// * `buf` must point to a valid, properly initialized `Buffer`.
+/// * `cursor` must point to a valid position within the Buffer's allocated memory.
+/// * The `BufferWriter` does not own the `Buffer`, only borrows it.
+/// * If the buffer is grown (implicitly) as part of a write operation, all previously obtained
+///   pointers into the buffer are invalidated.
+///
+/// # Usage
+///
+/// It's best to use this through the `std::io::Write` implementation, which handles
+/// buffer growth and cursor position management automatically.
+///
+/// ```no_run
+/// # use std::io::Write;
+/// # use buffer::{Buffer, BufferWriter};
+/// # use std::ptr::NonNull;
+/// # let buffer_ptr: NonNull<Buffer> = todo!();
+/// # let buffer = unsafe { buffer_ptr.as_mut() };
+/// let mut writer = BufferWriter {
+///     buf: buffer_ptr,
+///     cursor: buffer.data,  // Start writing at the beginning
+/// };
+///
+/// // Write data to the buffer
+/// writer.write(b"Hello, world!").unwrap();
+/// ```
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+pub struct BufferWriter {
+    pub buf: NonNull<Buffer>,
+    pub cursor: NonNull<u8>,
+}
+
+impl std::io::Write for BufferWriter {
+    fn write(&mut self, bytes: &[u8]) -> std::io::Result<usize> {
+        // Safety: We assume `buf` is a valid pointer to a properly initialized Buffer.
+        let buffer = unsafe { self.buf.as_mut() };
+
+        // Check if we need to grow the buffer to accommodate the new data
+        if buffer.len + bytes.len() > buffer.capacity {
+            // Safety: `Buffer_Grow` is a C function that increases the buffer's capacity. It
+            // expects a valid buffer pointer and returns the number of bytes added to capacity. If
+            // it returns non-zero, the buffer was successfully grown.
+            //
+            // Note that this may invalidate previously held pointers into the buffer.
+            if unsafe { Buffer_Grow(self.buf, bytes.len()) } != 0 {
```

Comment:
Shouldn't we handle the error case here?

Reply:
Oh, actually `0` doesn't mean an error. The only reason `Buffer_Grow` can fail is OOM I think and I don't think we handle that as an error anywhere (at least not in C). The docs are misleading. I'll fix them..

Reply:
Looked again at the varint C code and remembered that it's not using `Grow` directly but through `Buffer_Reserve` (an inline function so we can't use directly in Rust) that indeed does return `0` as an error if `buffer.len + bytes.len() > buffer.capacity`. However, `Grow` function doesn't return 0. I'll turn this into a debug assertion.